### ISSUE TITLE
Inject into Workers (androidx.WorkManager API)

### ISSUE THREAD

**Issue by droidrcc** [google/dagger#1183]
Hi,
from latest Google I/O you just presented the new `androidx.work.Worker` class, part of the new [WorkManager API](https://developer.android.com/topic/libraries/architecture/workmanager). 

Since the `Worker` class is created by the framework (we only pass the Worker class type to the `WorkManager`), how can we @inject fields into the Worker ? 
Do you intend to add a new AndroidInjection.inject() function that takes a worker as argument ?

Thanks

**Comment by bopbi** [comment]
in my case i inject it using inside doWork()

```
if (applicationContext is MyApp) {
(applicationContext as MyApp).userComponent.inject(this)

....
}
```

*MyApp is the application class name for my android app

**Comment by autonomousapps** [comment]
I know this is the wrong repo for this comment, but I wonder aloud why they couldn't have created a WorkerFactory class similar to ViewModelProvider.Factory for defining how to create Worker objects. Or something. I just really don't understand why Android framework classes can never have useful public constructors.

**Comment by andrew-veresov** [comment]
Hope they'll add it soon

**Comment by kuno** [comment]
@ronshapiro 

hi there, do you have the roughly release date for this `dagger-android-work` feature ?

**Comment by ronshapiro** [comment]
There are a few things that need to change, let me document them here:
1. [ ] Add `@ContributesAndroidInjector` support
2. [ ] Add an `AndroidWorkInjection.inject()` method, and probably also a map key.
3. [ ] Figure out a good way to integrate both for `DaggerApplication`. The `androidx` effort is fantastic in that it splits up a lot of things into individual libraries, but it complicates the approach we've taken thus far with `DaggerApplication`. Specifically, someone could want a `DaggerApplication` that has `androidx.work` support but not `androidx.fragment` support, and we shouldn't drag in any of those dependencies if we don't need to. This is reliant on the first two items, but also requires some thought about how to support the different flavors of android injection that we want/may want in the future. Codegen could help here too, but it may not be the best option.

We also probably want to wait until the library is at least in beta, or at least we would have to release an "alpha" version of dagger-work.

Lastly, we need to modify our bazel configuration to pull in the android-work aars, hopefully that won't be too hard.

**Comment by ferrerojosh** [comment]
I made a gist of how I injected workers, its mostly the same as how other components are injected

https://gist.github.com/ferrerojosh/82bd92748f315155fa6a842f4ed64c82

**Comment by samybeng** [comment]
Hello guys,

Thanks to @ferrerojosh !

I made a Java version for that in case some people need it.
https://gist.github.com/ThePredators/1702e79c5d3860f415c542da446420eb

**Comment by something15525** [comment]
Any updates on this now that androidx is released?

**Comment by ronshapiro** [comment]
The Workers API is still in alpha

**Comment by something15525** [comment]
Ah, good point.

**Comment by usmanrana07** [comment]
I got throw a solution that's working in rather simple way. How good is that? Is there any pros/cons of it over the above discussed solutions (though above discussed ones didn't work for me). 
Answer: https://stackoverflow.com/a/52199645/5110536

**Comment by sariego** [comment]
WorkManager-alpha9 introduced WorkerFactory, which would allow you to use constructor injection with Dagger :D
WorkManager-alpha10 introduced breaking changes, one of which is 
> The interface WorkerFactory and the concrete implementation DefaultWorkerFactory have been merged into an abstract class called WorkerFactory

[source](https://developer.android.com/jetpack/docs/release-notes)

**Comment by CDRussell** [comment]
Using `WorkerFactory` certainly seems like the right way going forward. I wrote up more of an explanation on how it is actually used - https://stackoverflow.com/a/53377279/1654145

**Comment by something15525** [comment]
Hey all,

Looks like WorkManager was release to beta today:
https://medium.com/androiddevelopers/introducing-workmanager-2083bcfc4712

From what I've read here: https://developer.android.com/jetpack/docs/release-notes#december_19_2018, it says:
`This release contains no API changes; moving forward, WorkManager is expected to stay API stable until the next version unless there is a critical problem.`

Does this mean no Dagger support is planning on being added? I realize this is probably not the ideal place to ask this question, but don't know where else to ask.

**Comment by ronshapiro** [comment]
That's good. We still have some other dagger.android architecture cleanups to get done before we can get this, but I hope soon. We've been spending most of our efforts on buildperf recently

**Comment by something15525** [comment]
@ronshapiro okay, good to know it will be gotten to eventually. Thanks for the update!

**Comment by arunkumar9t2** [comment]
@ronshapiro I just finished setting up constructor injection using multibinding for `Worker`s using [WorkerFactory](https://developer.android.com/reference/androidx/work/WorkerFactory). 

Given the possibility of constructor injection, should members injection via `AndroidInjection.inject(this)` still be considered/encouraged?

I wrote about my implementation [here](https://www.arunkumarsampath.in/dagger-recipes-illustrative-step-by-step-guide-to-achieve-constructor-injection-in-workmanager/). Full working sample [here](https://github.com/arunkumar9t2/dagger-workmanager).

**Comment by JakeWharton** [comment]
I'd say members injection is always inferior to constructor injection.
Fragments can be constructor injected as of 1.1+. Activities on 28+.

**Comment by raiytu4** [comment]
thank to Jake's [presentation](https://jakewharton.com/helping-dagger-help-you/) 

we now able to perform constructor injection, like so:

```
class HelloWorldWorker @AssistedInject constructor(
    @Assisted private val params: WorkerParameters,
    private val appContext: Context,
    private val foo: Foo
) : Worker(appContext, params) {
    private val TAG = "HelloWorldWorker"
    override fun doWork(): Result {
        Log.d(TAG, "Hello world!")
        Log.d(TAG, "Injected foo: $foo")
        return Result.success()
    }

    @AssistedInject.Factory
    interface Factory  : ChildWorkerFactory<HelloWorldWorker>
}
```
no subcomponent needed

fully working sample can be found here
https://github.com/raiytu4/dagger-workmanager

**Comment by Ethan1983** [comment]
@raiytu4 If going with AssistedInjection, it would be nice to have a @WorkerInject similar to @InflationInject to avoid this layers of factories. Ideally, it would be nice for WorkManager library to provide these and that goes to what Jake was talking about Dagger providing customization for libraries to provide modules.

**Comment by chrisbanes** [comment]
I've (finally) just migrated to WM v1.0.0-beta01. Thanks to @raiytu4 for the nice example.

You can simplify it even further by just using the `Context` which you're given in the Factory. I did that in: https://github.com/chrisbanes/tivi/pull/254

**Comment by arunkumar9t2** [comment]
Seems `AssistedInject` is easier compared to my subcomponent + multibinds approach. 

Any reason the `Context` could be anything other than the Application context? I thought this was a reasonable assumption to make, so ended up providing it from my root app component. The only param known at runtime was the `workerParameters`.

**Comment by nlgtuankiet** [comment]
@arunkumar9t2 You can deep dive into the WorkManager source code to find out, from what I see is that the `Context` parameter is always your Application context. 

@chrisbanes I checked out your PR, nice improvement you got there (remove the generic T), thank
```kotlin
interface ChildWorkerFactory {
    fun create(context: Context, params: WorkerParameters): ListenableWorker
}
```

But, I don't get the part

> You can simplify it even further by just using the `Context` which you're given in the Factory.

About the context, I basically do the same thing as you did in the beginning, then I really that the context is just our application context. And the point is, we only need "assisted" for parameters that can only be known at runtime. So there is no point to "assisted" the context.

**Comment by chrisbanes** [comment]
I disagree, if a factory passes you a parameter you use it, rather than what you think it will pass you.

**Comment by Ethan1983** [comment]
@chrisbanes Any reason not to have TiviWorkerFactory as singleton?

**Comment by iamjoin7** [comment]
> I've (finally) just migrated to WM v1.0.0-beta01. Thanks to @raiytu4 for the nice example.
> 
> You can simplify it even further by just using the `Context` which you're given in the Factory. I did that in: [chrisbanes/tivi#254](https://github.com/chrisbanes/tivi/pull/254)

Hey @chrisbanes. Thank you a lot of the example. I've been dealing with a few hours to set my WorkManager up like yours and I've done so but I'm stuck at a very weird bug. Would you have any idea on why this would happen?

When I do:
`workManager.enqueue(wWorkRequest)` 

The Work runs (doWork() is called)

but when I do:
`workManager.enqueueUniqueWork("Name", ExistingWorkPolicy.APPEND, wWorkRequest)` 

the work doesn't get start (doWork() is never called).

Prior to all my changes today such as AppWorkerFactory, ChildWorkerFactory, and all the changes you did in https://github.com/chrisbanes/tivi/pull/254, my enqueueUniqueWork used to work. So now injecting to the Worker classes via the way you've shown, somehow enqueueUniqueWork doesn't work. Only enqueue works.

Any help is appreciated! Thank you for your or anyone's time who can help out.

**Comment by chrisbanes** [comment]
@iamjoin7 Weird. Have you stepped through the WorkerFactory to see it's `createWorker()` is called?

**Comment by iamjoin7** [comment]
@chrisbanes yes I had. `createWorker()` actually does _not_ get called.

Although, after inspecting it more, this may be a WorkManager issue. I've checked the db with `stetho` and it seems the works are actually being saved, but never "run"ed

https://issuetracker.google.com/issues/121436946

It's quite weird how `enqueueUniqueWork` was working injecting via `AndroidWorkerInjector.inject(this)` but it's not via this `@AssistedInject.Factory` 🤔

Thank you for your time though @chrisbanes. I'm leaving my comment up in case someone else faces a simiar issue.

**Comment by nlgtuankiet** [comment]
@iamjoin7 I tested it out, work fine on my part though, could you provide more detail?
https://github.com/nlgtuankiet/dagger-workmanager/blob/iamjoin7-issue/app/src/main/java/com/sample/daggerworkmanagersample/MainActivity.kt

**Comment by vincent-paing** [comment]
The worker api is stable now, will something like `@ContributesAndroidInjector` be introduced in Dagger

**Comment by arunkumar9t2** [comment]
There are already two working approaches discussed in this thread and both involve constructor injection instead of members injection. One is by using subcomponents and multibindings ([link](https://github.com/arunkumar9t2/dagger-workmanager)) and other one is using Square's [AssistedInject](https://github.com/square/AssistedInject) ([link](https://github.com/nlgtuankiet/dagger-workmanager)).

IMHO, I don't think a `@ContributesAndroidInjector` and `AndroidInjection.inject(this)` solution is warranted anymore but I agree a working solution with simpler setup would be nice.

**Comment by DavidJRobertson** [comment]
The fact that the above solutions require changing the configuration of the (singleton) WorkManager instance makes them unusable in libraries.

**Comment by jega-ms** [comment]
here is example  for using worker and dagger. 
https://github.com/jega-ms/android-dagger2-mvp-rx/tree/master/app/src/main/java/com/jega/android/example/dagger2/worker

**Comment by hoanglm4** [comment]
+1

**Comment by heitorpr** [comment]
+1

**Comment by hoangdat** [comment]
@nlgtuankiet can you share me a example of test for your worker in this way?. Many thanks!!!

**Comment by nlgtuankiet** [comment]
@hoangdat 
https://proandroiddev.com/dagger-2-setup-with-workmanager-a-complete-step-by-step-guild-bb9f474bde37?source=friends_link&sk=f616f5361147403a793fa16fb378e071

**Comment by hoangdat** [comment]
> https://proandroiddev.com/dagger-2-setup-with-workmanager-a-complete-step-by-step-guild-bb9f474bde37?source=friends_link&sk=f616f5361147403a793fa16fb378e071

Srr, I did not find any test suite on it.

**Comment by Zhuinden** [comment]
> Given the possibility of constructor injection, should members injection via AndroidInjection.inject(this) still be considered/encouraged?

It's possible to use `AndroidInjection.inject(this)` (sort of) over `Worker`s since 2.20, where they removed the limitation that `AndroidInjector` was only generated for Activity/Fragment/Service/ContentProvider/BroadcastReceiver.

Now you can do it as simple as

``` kotlin
    @ContributesAndroidInjector
    abstract fun syncWorker(): SyncWorker
```

and

``` kotlin
    @Keep
    class SyncWorker(context: Context, params: WorkerParameters) : Worker(context, params) {
        init {
            val injector = context.applicationContext as HasAndroidInjector
            injector.androidInjector().inject(this)
        }
        @Inject
        lateinit var apiService: ApiService
```

**Comment by Chang-Eric** [comment]
This should be covered now with the Jetpack Hilt extension (https://developer.android.com/training/dependency-injection/hilt-jetpack#workmanager). Unfortunately for dagger.android, we likely won't be adding WorkManager support directly for that so I'm going to close this.
